### DISCUSSION TITLE
windows: Unable to build micropython STM32 port

### DISCUSSION THREAD

**Discussion by MOWA2** [micropython/micropython#16507] [STM32 / Pyboard]
Hi all,

I hope i'm posting on the right forum.

I am facing an issue when building micropython stm32 port on Windows. Problem seems to appear when linking.

Here is my environment:
- make (mingw32-make) 4.4.1
- gcc-arm-none-eabi 14.2
- python 3
Building on cygwin (tried MSYS2 too).

Everything seems to build fine, but I get this error (example when building for NUCLEO_F091RC:

```
CC sdram.c
CC fatfs_port.c
CC lcd.c
CC accel.c
CC servo.c
CC dac.c
CC adc.c
CC sdio.c
CC subghz.c
CC build-NUCLEO_F091RC/pins_NUCLEO_F091RC.c
LINK build-NUCLEO_F091RC/firmware.elf
c:/00-ws/99-tools/compil/arm-toolchain14.2/bin/../lib/gcc/arm-none-eabi/14.2.1/../../../../arm-none-eabi/bin/ld.exe: cannot find build-NUCLEO_F091RC/lib/libm/ef_rem_pi: No such file or directory
collect2.exe: error: ld returned 1 exit status
mingw32-make: *** [Makefile:614: build-NUCLEO_F091RC/firmware.elf] Error 1
```

I double checked, file is compiled, but it's name is "ef_rem_pio2".

When running with "V=1", I can see the link command. If I run it manually after failed build, and then I build the project again, it seems to work.

Does anyone know how to overcome this issue ?

Thank you.

**Comment by Josverl** [comment]
Im an avid Windows user, but wrt to building MicroPython I favour WSL2 and add docker desktop + mpbuild for simplicity

**Reply by MOWA2** [reply]
Yep, I decided to go the wsl route.

But I'm still searching for a solution, work laptop are running base windows without the wsl.

**Reply by Josverl** [reply]
If no WSL, then GitHub codespaces offers a free linux dev space for some 60? Cpu Hours a month, docker enabled. So install mpbuild and off you go.
Just need to download the built fw to local pc once its done

https://github.com/features/codespaces

On a corp laptop all other options are likely blocked.

**Reply by MOWA2** [reply]
Alright, thanks for the advice, seems like a good solution :)

**Comment by stinos** [comment]
> but it's name is "ef_rem_pio2".

Possibly the last 2 characters got trunctaed due to commandline length limits somewhere. IIRC I could work around that by creating symlinks for some of the directories. That was with MSYS2 though, which is actually nicer to work with than Cygwin imo (wrt package management, available packages, handling of paths).

**Reply by MOWA2** [reply]
Yeah, I thought the same for the length limits.. I tried plenty of solutions (changing build directory name, moving MP base directory at the root of C:, ...) but none of these worked unfortunately..

I can't see where it is blocking, or where the command line gets truncated

**Comment by stinos** [comment]
Forgot about this; I remembered having tried this before so I tried it again and it 'just' works: I can build stm32 in am MSYS2 shell just fine. So starting from an MSYS2 shell with git/python/make etc already installed, install the arm tools with `pacman -S mingw-w64-x86_64-arm-none-eabi-gcc` then build with `make CROSS_COMPILE=arm-none-eabi- BOARD=NUCLEO_F091RC`.
